### PR TITLE
Handle import scope dependencies in dependencyManagement (multiple goals) (Fixes multiple issues #308, #309, #346, #395)

### LINKED ISSUE #308: versions:use-dep-version does not support "import" scoped poms
Apache Maven 3.5.2
versions-maven-plugin 2.7

When using `versions:use-dep-version` the version of import scoped poms in the dependency management is not changed.
E.g.:
```xml
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.example.demo</groupId>
                <artifactId>aggregator</artifactId>
                <version>0.1.2-SNAPSHOT</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
</dependencyManagement>
```

### LINKED ISSUE #346: use-latest-releases ignores import scoped BOM dependencies
use-latest-releases ignores import scoped BOM dependencies

Apache Maven 3.6.0
versions-maven-plugin 2.7

example:

```
     <dependencyManagement>
         <dependencies>
             <dependency>
                 <groupId>org.springframework</groupId>
                 <artifactId>spring-framework-bom</artifactId>
                 <version>4.0.1.RELEASE</version>
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>
         </dependencies>
     </dependencyManagement>
 
     <dependencies>
         <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-context</artifactId>
         </dependency>
         <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-web</artifactId>
         </dependency>
     </dependencies>
```

Expected result: spring-framework-bom managed version to be updated (5.1.6.RELEASE at time of writing)
Actual result: no change to pom.xml

### LINKED ISSUE #395: use-dep-version (and other goals) does not handle the import-scoped dependency
* `versions:use-dep-version` does not update the version of the import-scoped dependency, in the `<dependencyManagement>` section. Other goals may also not handle import-scoped dependency correctly.
* This issue might be related to [Include import-scoped dependency in dependency plugin goals](https://issues.apache.org/jira/browse/MDEP-694)

### PR DESCRIPTION
**Pull request by frode-carlsen**
Fixes #308, Fixes #309, Fixes #346, Fixes #395 for following goals
* use-dep-version
* use-latest-versions
* use-latest-snapshots
* use-latest-release
* use-next-releases
* use-next-versions
* use-next-snapshots

### CODE REVIEW

File: src/main/java/org/codehaus/mojo/versions/UseNextVersionsMojo.java [hunk 1]
```diff
@@ -60,7 +60,11 @@ protected void update( ModifiedPomXMLEventReader pom )
         {
             if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
             {
+            	// handle normally declared dependencies
                 useNextVersions( pom, getProject().getDependencyManagement().getDependencies() );
+                
+                // handle import dependencies specially
+                useNextVersions( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseNextVersionsMojo.java:67]:
```suggestion
                useNextVersions( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseNextSnapshotsMojo.java [hunk 2]
```diff
@@ -97,7 +97,11 @@ protected void update( ModifiedPomXMLEventReader pom )
         {
             if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
             {
+            	// handle normally declared dependencies
                 useNextSnapshots( pom, getProject().getDependencyManagement().getDependencies() );
+                
+                // handle import dependencies specially
+                useNextSnapshots( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseNextSnapshotsMojo.java:104]:
```suggestion
                useNextSnapshots( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseNextReleasesMojo.java [hunk 3]
```diff
@@ -67,7 +67,11 @@ protected void update( ModifiedPomXMLEventReader pom )
     {
         if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
         {
+        	// handle normally declared dependencies
             useNextReleases( pom, getProject().getDependencyManagement().getDependencies() );
+            
+            // handle import dependencies specially
+            useNextReleases( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseNextReleasesMojo.java:74]:
```suggestion
            useNextReleases( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseLatestVersionsMojo.java [hunk 4]
```diff
@@ -96,7 +96,11 @@ protected void update( ModifiedPomXMLEventReader pom )
                     PomHelper.getRawModel( getProject() ).getDependencyManagement();
                 if ( dependencyManagement != null )
                 {
+                	// handle normally declared dependencies
                     useLatestVersions( pom, dependencyManagement.getDependencies() );
+                    
+                    // handle import dependencies specially
+                    useLatestVersions( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseLatestVersionsMojo.java:103]:
```suggestion
                    useLatestVersions( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseLatestSnapshotsMojo.java [hunk 5]
```diff
@@ -100,7 +100,11 @@ protected void update( ModifiedPomXMLEventReader pom )
         {
             if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
             {
+            	// handle normally declared dependencies
                 useLatestSnapshots( pom, getProject().getDependencyManagement().getDependencies() );
+                
+                // handle import dependencies specially
+                useLatestSnapshots( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseLatestSnapshotsMojo.java:107]:
```suggestion
                useLatestSnapshots( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseLatestReleasesMojo.java [hunk 6]
```diff
@@ -101,7 +101,11 @@ protected void update( ModifiedPomXMLEventReader pom )
         {
             if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
             {
+            	// handle normally declared dependencies
                 useLatestReleases( pom, getProject().getDependencyManagement().getDependencies() );
+                
+                // handle import dependencies specially
+                useLatestReleases( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseLatestReleasesMojo.java:108]:
```suggestion
                useLatestReleases( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

File: src/main/java/org/codehaus/mojo/versions/UseDepVersionMojo.java [hunk 7]
```diff
@@ -76,7 +76,11 @@ protected void update( ModifiedPomXMLEventReader pom )
         {
             if ( getProject().getDependencyManagement() != null && isProcessingDependencyManagement() )
             {
+            	// handle normally declared dependencies
                 useDepVersion( pom, getProject().getDependencyManagement().getDependencies() );
+                
+                // handle import dependencies specially
+                useDepVersion( pom, PomHelper.readImportedPOMsFromDependencyManagementSection(pom) );
```
Comment (mfriedenhagen) [inline src/main/java/org/codehaus/mojo/versions/UseDepVersionMojo.java:83]:
```suggestion
                // handle normally declared dependencies
                useDepVersion( pom, getProject().getDependencyManagement().getDependencies() );

                // handle import dependencies specially
                useDepVersion( pom, PomHelper.readImportedPOMsFromDependencyManagementSection( pom ) );
```

**Review by mfriedenhagen — CHANGES_REQUESTED** [review]
Hi @frode-carlson, 

first of all, thanks for this contribution. 
I have added remarks about formatting and it looks like you introduced tabs at a few places if I am not mistaken (before comment `// handle normally declared dependencies`). And the blank lines before `// handle import dependencies specially`
It would be mighty helpful if you included an invoker test as well which checks at least some of your changes.

**Comment by frode-carlsen** [comment]
added test case and applied fixes for tabs

**Comment by frode-carlsen** [comment]
@mfriedenhagen added test case and applied fixes for tabs

**Comment by nicktindall** [comment]
This would be really handy

**Comment by adpriebe** [comment]
Is there any chance of this PR getting merged? The documentation is very clear that the `dependencyManagement` section is processed by default, and therefore it is surprising that something like `mvn versions:use-latest-releases` does not also update import scoped dependencies in `dependencyManagement` to the latest release versions.

**Comment by ajarmoniuk** [comment]
It doesn't? Yesterday I checked it and it did update those. I did check that with maven 3.2.5 as well as 3.6.3.

Could you please share a test scenario where it doesn't update those dependencies? By that I understand a test pom.xml, not just a maven command line.

**Comment by adpriebe** [comment]
Hi @ajarmoniuk thanks for your fast response! Here is a contrived `pom.xml` that reproduces the issue:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>demo</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>demo</name>
	<description>Demo project for Spring Boot</description>
	<properties>
		<java.version>11</java.version>
	</properties>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-dependencies</artifactId>
				<version>2.7.0</version> <!-- old version -->
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter</artifactId>
			<scope>provided</scope>
		</dependency>
	</dependencies>
</project>
```

I have tested the above with both Maven 3.6.0 and 3.8.6 using command `mvn versions:use-latest-release -Dincludes=org.springframework.boot:spring-boot-dependencies`. The command succeeds without updating anything.

I would also note that the following `pom.xml` and command `mvn versions:update-properties -Dincludes=org.springframework.boot:spring-boot-dependencies` works as expected, with the `spring.version` property being updated:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>demo</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>demo</name>
	<description>Demo project for Spring Boot</description>
	<properties>
		<java.version>11</java.version>
                <spring.version>2.7.0</spring.version> <!-- old version -->
	</properties>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-dependencies</artifactId>
				<version>${spring.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter</artifactId>
			<scope>provided</scope>
		</dependency>
	</dependencies>
</project>
```

**Comment by ajarmoniuk** [comment]
Indeed, I was able to reproduce the issue using your POMs. Thank you. I will now update all mojos to include the fix for this issue.

**Comment by ajarmoniuk** [comment]
For some reason, I wasn't able to create a test using a bom in the `it-repo` repository, even if I copied the Spring Boot BOM there. Neither did using `org.codehaus.mojo:mojo-parent` reproduce the problem. In both cases, the BOM was present in `project.dependencyManagement.dependencies`.

**Pull request closed without merging**